From: Li Wang <liwang@redhat.com>
Subject: selftests/mm: serialize local_config generation for parallel builds
Date: Tue, 31 Mar 2026 12:01:54 +0800

The mm selftests generate both local_config.mk and local_config.h from
check_config.sh.  With high parallelism (-jN), this can race and lead to
incomplete target builds (e.g.  only a few binaries get built after 'make
-j100', while 'make -j1' builds everything).

Switch to a stamp-based dependency:

  local_config.stamp: check_config.sh
        ... run check_config.sh ...
        touch local_config.stamp

and make local_config.mk/local_config.h depend on the stamp.

This ensures check_config.sh is executed once per update decision and
removes the parallel race window.  Also hook local_config.stamp into
EXTRA_CLEAN.

No functional change intended for non-parallel builds.

Link: https://lkml.kernel.org/r/20260331040156.119158-2-liwang@redhat.com
Signed-off-by: Li Wang <liwang@redhat.com>
Reported-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 tools/testing/selftests/mm/Makefile |   10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

--- a/tools/testing/selftests/mm/Makefile~selftests-mm-serialize-local_config-generation-for-parallel-builds
+++ a/tools/testing/selftests/mm/Makefile
@@ -262,10 +262,14 @@ $(OUTPUT)/migration: LDLIBS += -lnuma
 
 $(OUTPUT)/rmap: LDLIBS += -lnuma
 
-local_config.mk local_config.h: check_config.sh
-	CC="$(CC)" CFLAGS="$(CFLAGS)" ./check_config.sh
+local_config.stamp: check_config.sh
+	$(call msg,CHK,config,$@)
+	$(Q)CC="$(CC)" CFLAGS="$(CFLAGS)" ./check_config.sh
+	$(Q)touch $@
 
-EXTRA_CLEAN += local_config.mk local_config.h
+local_config.mk local_config.h: local_config.stamp
+
+EXTRA_CLEAN += local_config.mk local_config.h local_config.stamp
 
 ifeq ($(IOURING_EXTRA_LIBS),)
 all: warn_missing_liburing
_
